### DISCUSSION TITLE
Recommended way to allow one tool installed with rye to find another tool installed with rye?

### DISCUSSION THREAD

**Discussion by gaia** [astral-sh/rye#509] [General]
On Debian Bookworm,

1) installed rye
2) `rye install urlwatch`
3) `rye install chump`

`chump `is an optional dependency (needed in my case) of `urlwatch`. `urlwatch` is called almost exclusively via root's crontab, but when I call it manually in the shell, to test `urlwatch`, it says `chump `is not installed. 

I don't think my situation is unique. How should this be handled? Should I use a project? Should I use some global something? What's the simplest way to help `urlwatch  find chump?

**Comment by zanieb** [comment]
I presume you're looking for `rye install urlwatch --extra-requirement chump` [docs](https://rye-up.com/guide/tools/#extra-requirements).

This is similar to [`pipx inject`](https://pypa.github.io/pipx/examples/#pipx-inject-example)

**Comment by gaia** [comment]
ok! that's as straightforward as it gets

but URLwatch doesn't seem to like it

```
Traceback (most recent call last):
  File "/root/.rye/tools/urlwatch/lib/python3.12/site-packages/urlwatch/handler.py", line 106, in process
    self.load()
  File "/root/.rye/tools/urlwatch/lib/python3.12/site-packages/urlwatch/handler.py", line 85, in load
    self.old_data, self.timestamp, self.tries, self.etag = self.cache_storage.load(self.job, guid)
                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/.rye/tools/urlwatch/lib/python3.12/site-packages/urlwatch/storage.py", line 571, in load
    for data, timestamp, tries, etag in CacheEntry.query(self.db,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 4, got 0)
```

I've been using urlwatch for years with the same config I have now, the only thing that changed was the upgrade to Debian Bookworm which caused me to not me able to run it via the usual method (apt install urlwatch && pip install chump, or pip install urlwatch && pip install chump). the urlwatch configuration files even are still the same from before the upgrade.

i understand if i should take this up with the urlwatch folks, but urlwatch has a pretty damn wide userbase and is fairly old (https://github.com/thp/urlwatch), so i'm not sure that's the right place to go.

note the error in question is only raised when there IS something to notify via chump. otherwise it works fine.

**Reply by zanieb** [reply]
Hm there are some similar issues over on the urlwatch repository, maybe related to migrating the cache? Can you share some verbose logs? You may also want to make sure the chump version is compatible, not sure what requirements they have.

**Comment by gaia** [comment]
turns out the error is intermittent, so let me try to narrow this down further before sending logs...

I already really appreciate your help, zanieb!

**Comment by gaia** [comment]
ok, here it is. forget chump, it's not related.

install urlwatch. setup a couple of jobs (`urlwatch --edit`)

```
name: "Test"
command: "date"
---
name: "Test 2"
command: "echo $HOSTNAME"
```

The first time you run it (`urlwatch -v`) it will show the jobs as new, as it should (they are new).
Now run it over and over and over until one of the jobs shows up as new again - it shouldn't, because it isn't new.

when chump is not installed and pushover is not set as a notification method, the error in the original post doesn't raise, you only get a polite notification for a new job which isn't new.

i am not sure this is related to rye, but given how many ppl use urlwatch, old jobs randomly alerting as if they were new would have a few issues on their side already...

PS: you might want to setup some notification, no need to be pushover, to match my environment. but i am not even sure this is needed. having a notification channel raises an error which should happen: the job is not new but it's being flagged as new.

**Comment by gaia** [comment]
installed urlwatch 2.25, again via rye, and the issue persisted. 2.25 works fine on other machines.

it is not an issue of some autoclean deleting the urlwatch cache - if it was then all jobs would alert as new. in this case only a subset of the jobs alerts as new. 

going to give up on rye for now. seemed great though...
